fedimint-server-core/src/migration.rs

```diff
@@ -0,0 +1,170 @@
+/// [`IServerDbMigrationContext`] injected by the `fedimint-server`.
+///
+/// Needs to be an extension trait, as `fedimint-server-core` can't
+/// implement things on general-purpose [`DbMigrationFnContext`]
+#[async_trait]
+pub trait ServerModuleDbMigrationFnContextExt<M>
+where
+    M: ServerModule,
+{
+    async fn get_typed_module_history_stream(
+        &mut self,
+    ) -> BoxStream<ModuleHistoryItem<<M as ServerModule>::Common>>;
+}
+
+#[async_trait]
+impl<M> ServerModuleDbMigrationFnContextExt<M> for ServerModuleDbMigrationFnContext<'_, M>
+where
+    M: ServerModule + Send + Sync,
+{
+    async fn get_typed_module_history_stream(
+        &mut self,
+    ) -> BoxStream<ModuleHistoryItem<<M as ServerModule>::Common>> {
+        let module_instance_id = self
+            .module_instance_id()
+            .expect("module_instance_id must be set");
+        let (dbtx, ctx) = self.split_dbtx_ctx();
+        Box::pin(
+            ctx
+                .ctx()
+                .get_module_history_stream(
+                    module_instance_id,
+                    dbtx
+                )
+                .await
+                .map(|item| match item {
+                    DynModuleHistoryItem::ConsensusItem(ci) => ModuleHistoryItem::ConsensusItem(
+                        ci.as_any()
+                            .downcast_ref::<<<M as ServerModule>::Common as ModuleCommon>::ConsensusItem>()
+                            .expect("Wrong module type")
+                            .clone(),
+                    ),
+                    DynModuleHistoryItem::Input(input) => ModuleHistoryItem::Input(
+                        input
+                            .as_any()
+                            .downcast_ref::<<<M as ServerModule>::Common as ModuleCommon>::Input>()
+                            .expect("Wrong module type")
+                            .clone(),
+                    ),
+                    DynModuleHistoryItem::Output(output) => ModuleHistoryItem::Output(
+                        output
+                            .as_any()
+                            .downcast_ref::<<<M as ServerModule>::Common as ModuleCommon>::Output>()
+                            .expect("Wrong module type")
+                            .clone(),
+                    ),
+                }),
+        )
+    }
+}
+
+/// A [`DbMigrationFn`] with inner-context type-specific for a given server
+/// module
+pub type ServerModuleDbMigrationFn<M> = DbMigrationFn<ServerModuleDbMigrationContext<M>>;
+
+/// A [`DbMigrationFn`] with inner-context type-erased for all server modules
+pub type DynServerDbMigrationFn = DbMigrationFn<DynServerDbMigrationContext>;
+/// A [`DbMigrationFnContext`] with inner-context type-erased around
+/// [`IServerDbMigrationContext`]
+pub type ServerDbMigrationFnContext<'tx> = DbMigrationFnContext<'tx, DynServerDbMigrationContext>;
+
+/// See [`apply_migrations_server_dbtx`]
+pub async fn apply_migrations_server(
```

Comment:
Server-side migrations now live in a server side crate. :partying_face: 

And it is lightweight `fedimint-server-core`.